support max_waiting_queries in workload

dorooleg opened the pull request:
Fixes https://github.com/ClickHouse/ClickHouse/issues/78697

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

The workload setting `max_waiting_queries` is now supported. It can be used to limit the size of the query queue. If the limit is reached, all subsequent queries will be terminated with the `SERVER_OVERLOADED` error.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

dorooleg committed:
first version for max_waiting_queries

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=81250) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=81250) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=81250) it.</sub>

pamarcos commented:
@dorooleg for some reason, you were attempting to merge against branch `query-scheduler` instead of master.
I changed that and fixed a conflict, but now there are many more changes.

Please take a look and check if everything is correct. You also need to write a proper description of what you do and why, leave only 1 entry in the Changelog category and write a new entry in the Changelog

dorooleg commented:
> @dorooleg for some reason, you were attempting to merge against branch `query-scheduler` instead of master. I changed that and fixed a conflict, but now there are many more changes.
> 
> Please take a look and check if everything is correct. You also need to write a proper description of what you do and why, leave only 1 entry in the Changelog category and write a new entry in the Changelog

Sorry, i published this merge request too early. Because it depends on `query-scheduler`. I will move to draft it. Also i changed the description like you said. In any way thank you that you check it so fast)

dorooleg committed:
test_max_waiting_queries has been added (first version)

dorooleg committed:
build has been fixed

dorooleg committed:
cleanup style

dorooleg committed:
syntex and exception handing have been fixed for tests

dorooleg committed:
queue_size and SchedulerNodeInfo

dorooleg committed:
max_waiting_query tests have been fixed

dorooleg committed:
max_waiting_queries doc

dorooleg committed:
mispell fixes

dorooleg committed:
limit check has been fixed

serxa commented on src/Common/Scheduler/ISchedulerNode.h:
```diff
@@ -79,6 +81,16 @@ struct SchedulerNodeInfo
+        if (value <= 0)
+            throw Exception(
+                ErrorCodes::INVALID_SCHEDULER_NODE,
+                "Zero, negative and non-finite node queue_size are not allowed: {}",
```
Let's make error message more user-friendly:
```suggestion
                "Workload setting `max_waiting_queries` value must be positive, got: {}",
```

serxa commented on src/Common/Scheduler/Nodes/FifoQueue.h:
```diff
@@ -59,6 +60,9 @@ class FifoQueue final : public ISchedulerQueue
             throw Exception(ErrorCodes::INVALID_SCHEDULER_NODE, "Scheduler queue is about to be destructed");
+
+        if (requests.size() >= static_cast<size_t>(info.queue_size))
+            throw Exception(ErrorCodes::SERVER_OVERLOADED, "Queue limit has been reached: {} of {}", requests.size(), info.queue_size);
```
Let's give a hit on what exact limit has been reached:
```suggestion
            throw Exception(ErrorCodes::SERVER_OVERLOADED, "Workload limit `max_waiting_queries` has been reached: {} of {}", requests.size(), info.queue_size);
```

serxa commented on src/Common/Scheduler/Nodes/UnifiedSchedulerNode.h:
```diff
@@ -263,8 +264,9 @@ class UnifiedSchedulerNode final : public ISchedulerNode
 
         // Should be called after constructor, before any other methods
-        [[nodiscard]] SchedulerNodePtr initialize(EventQueue * event_queue_)
+        [[nodiscard]] SchedulerNodePtr initialize(EventQueue * event_queue_, const WorkloadSettings& settings_)
```
Minor. Style
```suggestion
        [[nodiscard]] SchedulerNodePtr initialize(EventQueue * event_queue_, const WorkloadSettings & settings_)
```

serxa commented on tests/integration/test_scheduler_query/test.py:
```diff
@@ -194,3 +202,37 @@ def ensure_workload_concurrency(workload, limit: int) -> None:
+        """
+    )
+
+    pool_all = QueryPool(6, "all")
```
I think it should pass even with 7 queries: 1 running and 6 waiting
```suggestion
    pool_all = QueryPool(7, "all")
```

serxa reviewed (commented):
Looks good! A few small changes are required.

dorooleg committed:
Update src/Common/Scheduler/Nodes/UnifiedSchedulerNode.h

Co-authored-by: Sergei Trifonov <svtrifonov@gmail.com>

dorooleg committed:
Update src/Common/Scheduler/ISchedulerNode.h

Co-authored-by: Sergei Trifonov <svtrifonov@gmail.com>

dorooleg committed:
Update tests/integration/test_scheduler_query/test.py

Co-authored-by: Sergei Trifonov <svtrifonov@gmail.com>

dorooleg committed:
Update src/Common/Scheduler/Nodes/FifoQueue.h

Co-authored-by: Sergei Trifonov <svtrifonov@gmail.com>

dorooleg committed:
error message has been changed

serxa commented:
@maxknv It seems like a new CI problem. 

> [Unit tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81250&sha=latest&name_0=PR&name_1=Unit+tests+%28tsan%29)		21:39:56
Job did not provide Result: job script bug, died CI runner or praktika bug

maxknv commented:
> @maxknv It seems like a new CI problem.
> 
> > [Unit tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=81250&sha=latest&name_0=PR&name_1=Unit+tests+%28tsan%29)		21:39:56
> > Job did not provide Result: job script bug, died CI runner or praktika bug

It was superslow for some reason and killed by GH timeout https://github.com/ClickHouse/ClickHouse/actions/runs/15690829562/job/44207325537?pr=81250

the only ci problem is that it didnt handle timeout properly. ill check it

serxa commented on docs/en/operations/workload-scheduling.md:
```diff
@@ -285,7 +285,9 @@ CREATE WORKLOAD all SETTINGS max_concurrent_queries = 100, max_queries_per_secon
-Workload settings `max_queries_per_second` and `max_burst_queries` limits number of queries for the workload with a token bucket throttler. It guarantees that during any time interval `T` no more than `max_queries_per_second * T + max_burst_queries` new queries will start execution.
+Workload settings `max_queries_per_second` and `max_burst_queries` limit number of queries for the workload with a token bucket throttler. It guarantees that during any time interval `T` no more than `max_queries_per_second * T + max_burst_queries` new queries will start execution.
+
+Workload setting `max_waiting_queries` limits number of waiting queries for the workload. When the limit is reached, the server returns an error `SERVER_OVERLOADED`
```
```suggestion
Workload setting `max_waiting_queries` limits number of waiting queries for the workload. When the limit is reached, the server returns an error `SERVER_OVERLOADED`.
```

serxa committed:
Update docs/en/operations/workload-scheduling.md

serxa commented:
> 02160_client_autocomplete_parse_query

https://github.com/clickhouse/clickhouse/issues/81893

serxa commented:
Failed tests are unrelated. I will set the mergeable check manually
